core/src/main/java/io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java

```diff
@@ -180,33 +179,38 @@ private boolean checkLimit() {
     }
 
     private void fetchNextFrame() {
-        do {
-            this.cursor = frameSequence.next();
-            if (cursor > -1) {
-                PageFrameReduceTask task = frameSequence.getTask(cursor);
-                LOG.debug()
-                        .$("collected [shard=").$(frameSequence.getShard())
-                        .$(", frameIndex=").$(task.getFrameIndex())
-                        .$(", frameCount=").$(frameSequence.getFrameCount())
-                        .$(", active=").$(frameSequence.isActive())
-                        .$(", cursor=").$(cursor)
-                        .I$();
-
-                this.allFramesActive &= frameSequence.isActive();
-                this.rows = task.getRows();
-                this.frameRowCount = rows.size();
-                this.frameIndex = task.getFrameIndex();
-                if (this.frameRowCount > 0 && frameSequence.isActive()) {
+        try {
+            do {
+                this.cursor = frameSequence.next();
+                if (cursor > -1) {
+                    PageFrameReduceTask task = frameSequence.getTask(cursor);
+                    LOG.debug()
+                            .$("collected [shard=").$(frameSequence.getShard())
+                            .$(", frameIndex=").$(task.getFrameIndex())
+                            .$(", frameCount=").$(frameSequence.getFrameCount())
+                            .$(", active=").$(frameSequence.isActive())
+                            .$(", cursor=").$(cursor)
+                            .I$();
+                    this.allFramesActive &= frameSequence.isActive();
+                    this.rows = task.getRows();
+                    this.frameRowCount = rows.size();
+                    this.frameIndex = task.getFrameIndex();
                     this.frameRowIndex = 0;
-                    record.setFrameIndex(task.getFrameIndex());
-                    break;
+                    if (this.frameRowCount > 0 && frameSequence.isActive()) {
+                        record.setFrameIndex(task.getFrameIndex());
+                        break;
+                    } else {
+                        this.frameRowCount = 0; // force reset frame size if frameSequence was canceled or failed
+                        collectCursor(false);
+                    }
                 } else {
-                    collectCursor(false);
+                    Os.pause();
                 }
-            } else {
-                Os.pause();
-            }
-        } while (this.frameIndex < frameLimit);
+            } while (this.frameIndex < frameLimit);
+        } catch (Throwable e) {
+            LOG.error().$(e).$();
+            throw CairoException.instance(0).put(exceptionMessage).setInterruption(true);
```

Comment:
Do you think we should be using the same `timeout, query aborted` error message in this scenario? It would only happen when reduce on the local task fails, so it's not a timeout, but something like unhandled error.

Reply:
The exception message can be changed, but it must be the same for local/remote task failure. 
Actually it's handled error, but probably `unexpected`.

Reply:
In case of a remote task failure we don't really know the reason behind it: it could be a timeout, or it could be an exception thrown from the `reduce()` call. In both cases we end up with the same `timeout, query aborted` message (ideally, we should handle this cases differently).